src/services/job-processor/message-storage.ts

```diff
@@ -0,0 +1,110 @@
+import { ExitMessage, ExitMessageWithMetadata } from './service.js'
+
+export type ValidatorIndex = string
+export type TTL = number
+
+export class MessageStorage {
+  private MAX_TTL: TTL = 10
+
+  // Time to live counter for each message
+  private messageTTL: Map<ValidatorIndex, TTL> = new Map<ValidatorIndex, TTL>()
+
+  private messagesMetadatas: Map<ValidatorIndex, ExitMessageWithMetadata> =
+    new Map<ValidatorIndex, ExitMessageWithMetadata>()
+
+  private addOrUpdateMessage(
+    messageWithMetadata: ExitMessageWithMetadata
+  ): number {
+    const hasMsg = this.messagesMetadatas.has(
+      String(messageWithMetadata.data.message.validator_index)
+    )
+
+    this.messagesMetadatas.set(
+      String(messageWithMetadata.data.message.validator_index),
+      messageWithMetadata
+    )
+
+    this.setTTL(messageWithMetadata.data.message.validator_index, this.MAX_TTL)
+
+    if (!hasMsg) {
+      // msg was appended
+      return 1
+    }
+
+    // msg was updated (overwritten)
+    return 0
+  }
+
+  private setTTL(validatorIndex: ValidatorIndex, ttl: TTL) {
+    this.messageTTL.set(String(validatorIndex), ttl)
+  }
+
+  public updateMessages(messagesWithMetadata: ExitMessageWithMetadata[]): {
+    updated: number
+    added: number
+  } {
+    return messagesWithMetadata
+      .map((msg) => this.addOrUpdateMessage(msg))
+      .reduce(
+        (stats, x) => {
+          if (x === 1) {
+            stats.added++
+            return stats
+          }
+          stats.updated++
+          return stats
+        },
+        { updated: 0, added: 0 }
+      )
+  }
+
+  public hasMessageWithChecksum(fileChecksum: string): boolean {
+    for (const msgWithMeta of this.messagesMetadatas.values()) {
+      if (msgWithMeta.meta.fileChecksum === fileChecksum) {
+        this.setTTL(msgWithMeta.data.message.validator_index, this.MAX_TTL)
+        return true
+      }
+    }
+
+    return false
+  }
```

Comment:
It's not obvious that a method named `hasSomething` and returning `boolean` mutates anything when called